### ISSUE TITLE
Move history to a new location 

### ISSUE THREAD

**Issue by wltjr** [rbtylee/e21-modules-extra#5]
More a feature request than an issue.

It seems that the clipboard module is storing its history file in the following location
```
~/cliboard/history
```

That seems a bit odd. It should likely be hidden at minimum. But may be best in one of the generic directories. I think offhand, maybe
```
~/.cache/clipboard
```

Or something like that. Configuration stuff tends to be under
```
~/.config
```

It really depends on if clipboard will have other uses for its own directory. If the only use is for the history file. Then cache might make the most sense. Really up to you. I just do not think it should be a visible top level directory in a users home directory. Not like they can really edit that file or do anything other than remove.

**Comment by rbtylee** [comment]
Look at the function **_set_data_path** in the file [history.c](https://github.com/rbtylee/e21-modules-extra/blob/master/Clipboard/src/history.c). The clipboard module tries to use the $XDG_DATA_HOME environmental variable to determine this location ($XDG_DATA_HOME), if that is not set then it uses $HOME/.local/share/clipboard and if $HOME is not set then it uses ./clipboard.

That is reasonable behavior I think. In bodhi/ubuntu the location is ~/.local/share/clipboard.

Check your setup and see if at least $HOME is defined or $XDG_DATA_HOME.

**Comment by wltjr** [comment]
Home is definitely set, have to check when the clipboard module is started. But being part of a login session, HOME should be set before E is even started.
```
$ printenv | grep HOME
JAVA_HOME=/etc/jem/vm
HOME=/home/wlt
JAVACC_HOME=/usr/share/javacc/
JDK_HOME=/etc/jem/vm
ANT_HOME=/usr/share/ant
```

But the location really should not change based on some env variable. I definitely have a ~/.local/share. but clipboard is not there. That should be the location always.

It may be worth considering something like the following, that does not rely on env vars. If that ends up being problematic. I know [Coverity reports getenv stuff as tainted](https://stackoverflow.com/q/27352216). Which is a total PITA to resolve, [example of attempt to fix](https://github.com/Obsidian-StudiosInc/jem/commit/011eb27f901107a858b29eff576e38bef71c0e8e) but did not.
```
#include <unistd.h>
#include <sys/types.h>
#include <pwd.h>

struct passwd *pw = getpwuid(getuid());

const char *homedir = pw->pw_dir;
```
 [Source]( https://stackoverflow.com/a/2910392 )

**Comment by wltjr** [comment]
Speaking of Coverity you may want to setup [Travis CI](https://travis-ci.org/) and also [Coverity](https://scan.coverity.com/) for your projects. Coverity is very invaluable to newer programmers. It is helping me catch mistakes even in old code that I ran through valgrind. Great learning tool!!! Most stuff in E git repos I believe is run through Coverity. That is how I found out about it.

You can look at my [entrance](https://github.com/Obsidian-StudiosInc/entrance/blob/wltjr/.travis.yml) and [ecrire](https://github.com/Obsidian-StudiosInc/ecrire/blob/master/.travis.yml) travis.yml files for the basics. If you need help let me know more than happy to help. Or if need be, I can create the travis.yml files, etc. But it is pretty straight forward looking at examples.

**Comment by rbtylee** [comment]
> But the location really should not change based on some env variable. I definitely have a ~/.local/share. but clipboard is not there. That should be the location always.

As I explained in IRC channel, the clipboard module is attempting to following the [XDG Base directory specifications](https://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html). Perhaps the code as it is doesn't do enough to error check or completely follow [gnome developer guidelines](https://developer.gnome.org/basedir-spec/) [_not really a gnome app but those are reasonable guidelines_] or isn't admittedly cross platform, but the location of the history data most assuredly **should** change depending on some _env variable_. This is the standard adhered to by gnome, gtk+, kde, qt and many apps and libraries to numerous to mention but for example: vlc, Chrome/Chromium, GStreamer, ...

> Home is definitely set, have to check when the clipboard module is started. But being part of a login session, HOME should be set before E is even started.

I never expected HOME to not be set and the code to fall thru to 
```
      else {
         // Should never happen
         strcpy(path, "");
         success = EINA_FALSE;
     }
```
But in your case that has happened :( I am unsure why. That doesn't happen in debian/ubuntu based distros I have tested this code on. You mentioned  (_IRC_) using sddm and entrance DMs, out of curiosity  I installed sddm on a bodhi base, and still ended up with a history data location of **~/.local/share/clipboard**. I am curious as to what your exact set up (Distro et al) is? I would think something is misconfigured or not right about it.

Regardless I can simplify the code here to not use the $HOME variable but instead something like: 
```
    struct passwd *pw = NULL;
    temp_str = getenv ("XDG_DATA_HOME");
    if (temp_str!=NULL) {
      strcpy(path, temp_str);
      // Ensure XDG_DATA_HOME terminates in '/'
      if (path[strlen(path)-1] != '/')
        strcat(path,"/");
      strcat(path, DATA_DIR);
    }
    /* XDG_DATA_HOME default */
    else {
      pw = getpwuid(getuid());
      temp_str = pw->pw_dir;
      // Hopefully unnecessary Safety check
      if (temp_str!=NULL)
         sprintf(path,"%s/.local/share/",temp_str);
      else {
         // Should never happen
         strcpy(path, "");
         success = EINA_FALSE;
     }
```
That should resolve your issue and ensure that unless $XDG_DATA_HOME is set, the location of said file ends up at **~/.local/share/clipboard**

**PLEASE VERIFY THE ABOVE CODE RESOLVES YOUR ISSUE** 

> I know Coverity reports getenv stuff as tainted.

You are much concerned much about making Coverity happy but I don't currently use Coverity. I have considered it and may give it a try in the future. So worrying about what may or may not be reported by Coverity is pretty much at the bottom of my list of priorities at the moment. But from my point of view [the getenv stuff is tainted thing is a non issue](https://stackoverflow.com/questions/21703826/tainted-string-in-c):

> Error:TAINTED_STRING is warning that (as far as Coverity can tell) some aspect of the behavior is influenced by some external input and that the external input is not examined for 'safeness' before it influences execution.
> In this particular example it would appear that Coverity is wrong ...  So a reasonable course of action would be to mark the error as a non-issue so that it is ignored on future runs.

If and when I decide to use Coverity, I will worry about it then if it throws an error on this code segment.

From my experience using pylint, pychecker, and pyflakes, these tools can be very valuable finding actual errors but they can also generate alot of noise that are not really issues at all. I like my python programs to be _pythonic_ , so I usually pay attention to what these tools report, but as is the case with all such code analysis tools you will get false positives as well as potentially whining about not adhering to some set of conventions. Most such conventions are reasonable but not all need to be strictly adhered to in order for your code to be safe, secure, and function correctly. I expect Coverity is the same.

> Speaking of Coverity you may want to setup Travis CI and also Coverity for your projects. Coverity is very invaluable to newer programmers. It is helping me catch mistakes even in old code that I ran through valgrind. Great learning tool!!! Most stuff in E git repos I believe is run through Coverity. That is how I found out about it.

E's use of Coverity is also how i found out about it. Admittedly probably a useful tool and I will probably get around to using it on some of my projects. (_Seems I have the day off work so may play with it latter_). On Travis CI, just causally looking at it it seems its build environment is Ubuntu Trusty base and that limits it usefulness to me. Bodhi currently is using 16.04 and hence I would need to build and test on a Ubuntu Xenial base. At least that is what I would prefer.

**Comment by wltjr** [comment]
That looks good, but more I look at the code. I think for some reason I am hitting the first conditional. Since the dir is ~/clipboard. That seems like the XDG_DATA_HOME maybe empty vs null not sure. I need to debug.

With regard to Coverity and getenv. I think their check is saying that most do not sanitize that data. Like running some check beyond a null check. Which if the data is to long or in the wrong format, chars when it should be numbers, etc. Then it could cause issues. Thus its usually good to sanitize anything you get with getenv.

In this case it may be a empty check in addition to the null check. I will have to test out. But it really seems like it is hitting the first conditional for me, and not even making it to the HOME variable. Thus even with your change. I may still hit the first one, and that location seems to match up.

**Comment by rbtylee** [comment]
I refactored the history stuff. Try my latest commit.

**Comment by wltjr** [comment]
Yes, good to go now. Not sure what fixed it, either the modification to the XDG conditional, adding !="/", or the correction to the use of the path variable. That may have caused the issue itself, and just seemed like it was hitting the first conditional. Since the 2nd had .local/share in the path, but if that got nuked. It was just using ~/.

I moved my ~/clipboard directory to ~/.local/share, when logged out. When I logged in, my history was there, and the directory was not in ~/clipboard, but wrote to the file in ~/.local/share/clipboard. Thus we should be good to go! Thank you for addressing that issue.
